File: src/main/java/beam/physsim/jdeqsim/akka/EventManagerActor.java
package beam.physsim.jdeqsim.akka;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import beam.utils.DebugLib;
import org.matsim.api.core.v01.events.Event;
import org.matsim.api.core.v01.network.Network;
import org.matsim.core.api.experimental.events.EventsManager;
import org.matsim.core.config.groups.TravelTimeCalculatorConfigGroup;
import org.matsim.core.events.EventsManagerImpl;
import org.matsim.core.events.handler.EventHandler;
import org.matsim.core.trafficmonitoring.TravelTimeCalculator;

import java.util.LinkedList;

public class EventManagerActor extends UntypedActor {

    ActorRef jdeqsimActorREF;
    public static final String LAST_MESSAGE = "lastMessage";
    public static final String REGISTER_JDEQSIM_REF = "registerJDEQSimREF";

    private TravelTimeCalculator travelTimeCalculator;
    private EventsManager eventsManager;
    private Network network;

    public EventManagerActor(Network network){
        this.network=network;
        resetEventsActor();
    }

    private void resetEventsActor(){
        eventsManager=new EventsManagerImpl();
        TravelTimeCalculatorConfigGroup ttccg = new TravelTimeCalculatorConfigGroup();
        travelTimeCalculator = new TravelTimeCalculator(network, ttccg);
        eventsManager.addHandler(travelTimeCalculator);
    }

    @Override
    public void onReceive(Object msg) throws Exception {
        if (msg instanceof Event) {
            eventsManager.processEvent((Event) msg);
        } else if (msg instanceof String) {
            String s = (String) msg;
            if (s.equalsIgnoreCase(LAST_MESSAGE)) {
                jdeqsimActorREF.tell(travelTimeCalculator, getSelf());
                resetEventsActor();
            } else if (s.equalsIgnoreCase(REGISTER_JDEQSIM_REF)) {
                jdeqsimActorREF = getSender();
            }else {
                DebugLib.stopSystemAndReportUnknownMessageType();
            }
        }  else {
            DebugLib.stopSystemAndReportUnknownMessageType();
        }

    }

    public static Props props(Network network) {
        return Props.create(EventManagerActor.class,network);
    }

}

File: src/main/scala/beam/physsim/jdeqsim/akka/EventManagerActor.scala
package beam.physsim.jdeqsim.akka

import java.io.File

import akka.actor.{Actor, ActorLogging, ActorRef, Props, Stash, UntypedActor}
import beam.utils.DebugLib
import org.matsim.api.core.v01.events.Event
import org.matsim.api.core.v01.network.Network
import org.matsim.core.api.experimental.events.EventsManager
import org.matsim.core.config.groups.TravelTimeCalculatorConfigGroup
import org.matsim.core.events.EventsManagerImpl
import org.matsim.core.events.handler.EventHandler
import org.matsim.core.trafficmonitoring.TravelTimeCalculator
import java.util

import akka.event.LoggingReceive
import beam.sim.BeamServices
import org.matsim.core.events.algorithms.EventWriterXML
import org.matsim.core.network.NetworkUtils


object EventManagerActor {
  val LAST_MESSAGE = "lastMessage"
  def REGISTER_JDEQSIM_REF = "registerJDEQSimREF"
  def props(beamServices: BeamServices): Props = Props.create(classOf[EventManagerActor], beamServices)
}

class EventManagerActor(var beamServices: BeamServices) extends Actor with Stash with ActorLogging {

  resetEventsActor()
  var jdeqsimActorREF: ActorRef = _
  var travelTimeCalculator: TravelTimeCalculator = _
  var eventsManager: EventsManager = _
  var eventsWriterXML: EventWriterXML = _

  private def resetEventsActor(): Unit = {
    eventsManager = new EventsManagerImpl
    val ttccg = new TravelTimeCalculatorConfigGroup
    travelTimeCalculator = new TravelTimeCalculator(beamServices.matsimServices.getScenario.getNetwork, ttccg)
    eventsManager.addHandler(travelTimeCalculator)
    addEventWriter
  }

  private def addEventWriter = {
    val writeEventsInterval = beamServices.beamConfig.beam.outputs.writePhysSimEventsInterval
    val iteration = beamServices.matsimServices.getIterationNumber

    if (writeEventsInterval == 1 || (writeEventsInterval > 0 && beamServices.matsimServices.getIterationNumber / writeEventsInterval == 0)) {
      createNetworkFile
      eventsWriterXML = new EventWriterXML(beamServices.matsimServices.getControlerIO.getIterationFilename(iteration, "physSimEvents.xml.gz"))
      eventsManager.addHandler(eventsWriterXML)
    }
  }

  private def createNetworkFile = {
    val physSimNetworkFilePath = beamServices.matsimServices.getControlerIO.getOutputFilename("physSimNetwork.xml.gz")
    if (!(new File(physSimNetworkFilePath)).exists()) {
      NetworkUtils.writeNetwork(beamServices.matsimServices.getScenario.getNetwork, physSimNetworkFilePath)
    }
  }

  override def receive = LoggingReceive{

    case event: Event => eventsManager.processEvent(event)
    case s: String => {
      if (s.equalsIgnoreCase(EventManagerActor.LAST_MESSAGE)) {
        jdeqsimActorREF.tell(travelTimeCalculator, self)
        if (eventsWriterXML!=null){
          eventsWriterXML.closeFile()
        }
      }
      else if (s.equalsIgnoreCase(EventManagerActor.REGISTER_JDEQSIM_REF)) jdeqsimActorREF = sender
      else DebugLib.stopSystemAndReportUnknownMessageType()
    }
    case _ => DebugLib.stopSystemAndReportUnknownMessageType()
  }
}
